src/transformStateWithClones.js

```diff
@@ -7,7 +7,27 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newState = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => delete stateCopy[key]);
+        break;
+
+      case 'clear':
+        stateCopy = {};
+        break;
+    }
+    newState.push({ ...stateCopy });
```

Comment:
Instead of creating a new copy of 'stateCopy' every time in the loop, consider moving the cloning of the state outside of the switch statement. This will follow the DRY (Don't Repeat Yourself) principle and reduce redundancy.